src/cmd/services/m3coordinator/downsample/downsampler.go

```diff
@@ -94,34 +100,90 @@ func newDownsampler(opts downsamplerOptions) (*downsampler, error) {
 	if logger.Check(zapcore.DebugLevel, "debug") != nil {
 		debugLogging = true
 	}
 
-	metricsAppenderOpts := metricsAppenderOptions{
-		agg:                          opts.agg.aggregator,
-		clientRemote:                 opts.agg.clientRemote,
-		defaultStagedMetadatasProtos: opts.agg.defaultStagedMetadatasProtos,
-		clockOpts:                    opts.agg.clockOpts,
-		tagEncoderPool:               opts.agg.pools.tagEncoderPool,
-		matcher:                      opts.agg.matcher,
-		metricTagsIteratorPool:       opts.agg.pools.metricTagsIteratorPool,
-		debugLogging:                 debugLogging,
-		logger:                       logger,
-		augmentM3Tags:                opts.agg.m3PrefixFilter,
+	return metricsAppenderOptions{
+		agg:                    agg.aggregator,
+		clientRemote:           agg.clientRemote,
+		clockOpts:              agg.clockOpts,
+		tagEncoderPool:         agg.pools.tagEncoderPool,
+		matcher:                agg.matcher,
+		metricTagsIteratorPool: agg.pools.metricTagsIteratorPool,
+		debugLogging:           debugLogging,
+		logger:                 logger,
+		augmentM3Tags:          agg.m3PrefixFilter,
 	}
-
-	return &downsampler{
-		opts:                opts.opts,
-		agg:                 opts.agg,
-		metricsAppenderOpts: metricsAppenderOpts,
-	}, nil
 }
 
 func (d *downsampler) NewMetricsAppender() (MetricsAppender, error) {
 	metricsAppender := d.agg.pools.metricsAppenderPool.Get()
+
+	d.RLock()
 	metricsAppender.reset(d.metricsAppenderOpts)
+	d.RUnlock()
+
 	return metricsAppender, nil
 }
+
+func (d *downsampler) OnUpdate(namespaces m3.ClusterNamespaces) {
+	logger := d.opts.InstrumentOptions.Logger()
+
+	if len(namespaces) == 0 {
+		logger.Debug("received empty list of namespaces. not updating staged metadata")
+		return
+	}
+
+	autoMappingRules, err := NewAutoMappingRules(namespaces)
+	if err != nil {
+		logger.Error("could not generate automapping rules for aggregated namespaces."+
+			" aggregations will continue with current configuration.", zap.Error(err))
+		return
+	}
+	var defaultStagedMetadatasProtos []metricpb.StagedMetadatas
+	for _, rule := range autoMappingRules {
+		metadatas, err := rule.StagedMetadatas()
+		if err != nil {
+			logger.Error("could not generate staged metadata from automapping rules."+
+				" aggregations will continue with current configuration.", zap.Error(err))
+			return
+		}
+
+		var metadatasProto metricpb.StagedMetadatas
+		if err := metadatas.ToProto(&metadatasProto); err != nil {
+			logger.Error("could not generate staged metadata from automapping rules."+
+				" aggregations will continue with current configuration.", zap.Error(err))
+			return
+		}
+
+		defaultStagedMetadatasProtos = append(defaultStagedMetadatasProtos, metadatasProto)
+	}
+
+	newMetricsAppenderOpts := defaultMetricsAppenderOptions(d.opts, d.agg)
+	newMetricsAppenderOpts.defaultStagedMetadatasProtos = defaultStagedMetadatasProtos
+
+	d.Lock()
+	d.metricsAppenderOpts = newMetricsAppenderOpts
+	d.Unlock()
```

Comment:
nit; I'm pretty sure this can be simplified to 
```suggestion
	d.Lock()
	d.metricsAppenderOpts.defaultStagedMetadatasProtos = defaultStagedMetadatasProtos
	d.Unlock()
```

Reply:
Overthought this when I was implementing. Think you're right. Will fix up